third_party/nvfuser/csrc/fusion_segmenter.cpp

```diff
@@ -3120,17 +3106,6 @@ void SegmentCandidateFinder::findSegments() {
 
   segmented_fusion_->validateIfDebug();
 
-  // All merges will be vertical beyond this point for now, so
-  //  we can remove the input auxiliary groups. Should make the vertical
-  //  merges avoid auxiliary group once we start general horizontal merges
-  std::unordered_set<SegmentedGroup*> input_groups;
-  for (auto input : forwarded_fusion_inputs_) {
-    input_groups.insert(input2group_.at(input));
-  }
-  eraseGroups(input_groups);
```

Comment:
This is no longer the case as insertion of type casting relies on edges, so the inputs need to be associated with edges. Erasing the input groups and their edges here causes failures to down cast inputs.

In the PR, the removal of the input groups is done just before the finalization step. This means that the input groups remain during the vertical merge step, but we don't want to consider them for merge, so those groups need to be explicitly tagged as "input" and be ignored. That's why `SegementedGroup` has a bool field, `is_fusion_input_`.